protoc-gen-twirp/generator.go

```diff
@@ -751,6 +764,31 @@ func (t *twirp) generateUtils() {
 	t.P(`  return h.Error(ctx, err)`)
 	t.P(`}`)
 	t.P()
+
+	t.P(`// BufferPool represents a pool of buffers. The *sync.Pool type satisfies this`)
+	t.P(`// interface.  The type of the value stored in a pool is not specified.`)
```

Comment:
Right, we have _internal_ expectations (in our case, our pooled objects are always `[]byte`), but the point of designing the API this way is so that we can provide an user-facing interface that will _not_ break in the future if for instance we decide to e.g. start pooling `bytes.Buffer` directly. I think requiring that the users make no assumptions about the data we're pooling is the only way to accomplish this.

FWIW this pattern was taken directly from how `gorilla/websocket` does write-buffer pooling, which I think works well in practice: https://godoc.org/github.com/gorilla/websocket#Dialer

On the other hand, if you think we should always be using memory pools by default, there is no point on keeping this interface.

Reply:
The PR's current code uses the buffer by doing this:
```go
if b := pool.Get(); b != nil {
  return b.([]byte)
}
```
which will panic if the `pool.Get()` returns _anything_ except a `[]byte`.

That means that all users of the library today would need to supply a `[]byte` pool. I don't think the current code lets users make no assumptions - it just makes their program crash if they guess wrong. It doesn't give us future flexibility, at least not as written.

The main value of the `interface{}` signature is that it lets users pass in a `sync.Pool` instead of making their own wrapper type, which is okay.

Reply:
But `sync.Pool` objects are not internally typed! :sweat_smile: 

We're asking the users to pass in a pool that _makes no assumptions about types_, meaning that when you call `pool.Get()`, if there are no objects in the pool, you don't return anything. This is the default behavior of `sync.Pool` unless you explicitly pass a constructor for new objects. The code has been designed to handle pools that don't construct objects -- we construct them ourselves in the codegen, so if we want to start constructing `proto.Bytes` objects, the same pool would work the same.

Does that make sense? Have you seen the `gorilla/websocket` docs to see how they implemented the same pattern? I think it worked well for them because they changed the underlying pool type once without breaking user code.